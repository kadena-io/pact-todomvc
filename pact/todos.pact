;;
;; todos smart contract
;;

;; admin keyset definition
(define-keyset 'todo-admin-keyset
  (read-keyset "todo-admin-keyset"))

;; todos module, administered by keyset
(module todos 'todo-admin-keyset
  " Smart contract module for TODO-MVC pact app.   \
  \ Tables:                                        \
  \ todo -- holds todo entries"

  ;; todo schema and table
  (defschema todo
    "Row type for todos."
     title:string
     completed:bool
     deleted:bool
     )

  (deftable todo-table:{todo})

  ;;
  ;; API functions
  ;;

  (defun new-todo (id title)
    "Create new todo with ENTRY and DATE."
    (insert todo-table id {
      "title": title,
      "completed": false,
      "deleted": false })
  )

  (defun toggle-todo-status (id:string)
    "Toggle completed status flag for todo at ID."
    (with-read todo-table id {
      "completed":= state
      }
      (update todo-table id {
        "completed": (not state) })))

  (defun edit-todo (id:string title)
    "Update todo ENTRY at ID."
    (update todo-table id {
      "title": title }))

  (defun delete-todo (id:string)
    "Delete todo title at ID (by setting deleted flag)."
    (update todo-table id {
      "deleted": true }))

  (defun read-todo:object (id:string)
    "Read a single todo"
    (+ {'id: id} (read todo-table id)))

  (defun read-todos:[object{todo}] ()
    "Read all todos."
    (map (read-todo) (keys todo-table)))
)

(create-table todo-table)
;done
