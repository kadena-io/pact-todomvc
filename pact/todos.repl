(env-keys ["todoadmin"])
(env-data { "todo-admin-keyset":
  { "keys": ["todoadmin"], "pred": "keys-all" } })

(begin-tx)
(load "todos.pact")
(commit-tx)

(typecheck "todos")

(use todos)

;;Create a new todo
(new-todo "1" "wash")

;;Test if todo was created
(expect "Returns the todo object with id-1" {"completed": false, "deleted": false, "id": "1", "title": "wash"} (read-todo "1"))

;;Fail if todo to be edited does not exist.
(expect-failure "No todo exists with id of 2." (edit-todo "2" "clean"))
