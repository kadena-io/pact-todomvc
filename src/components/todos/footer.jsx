import * as React from "react";

export class TodoFooter extends React.PureComponent {
  render() {
    var pluralize = function(count, word) {
      return count === 1 ? word : word + "s";
    };
    var activeTodoWord = pluralize(this.props.count, "item");
    var clearButton = null;

    if (this.props.completedCount > 0) {
      clearButton = (
        <button
          className="clear-completed"
          onClick={this.props.onClearCompleted}
        >
          Clear completed
        </button>
      );
    }

    var nowShowing = this.props.nowShowing;
    return (
      <footer className="footer">
        <span className="todo-count">
          <strong>{this.props.count}</strong> {activeTodoWord} left
        </span>
        <ul className="filters">
          <li>
            <a href="#/" className={{ selected: nowShowing === "all" }}>
              All
            </a>
          </li>{" "}
          <li>
            <a
              href="#/active"
              className={{ selected: nowShowing === "active" }}
            >
              Active
            </a>
          </li>{" "}
          <li>
            <a
              href="#/completed"
              className={{ selected: nowShowing === "completed" }}
            >
              Completed
            </a>
          </li>
        </ul>
        {clearButton}
      </footer>
    );
  }
}
