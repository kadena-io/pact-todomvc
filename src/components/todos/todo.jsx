import * as React from "react";

var ESCAPE_KEY = 27;
var ENTER_KEY = 13;

export class TodoItem extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = { editText: this.props.todo.title };
  }

  handleSubmit(event) {
    var val = this.state.editText.trim();
    if (val) {
      this.props.onSave(val);
      this.setState({ editText: val });
    } else {
      this.props.onDestroy();
    }
  }

  handleEdit() {
    this.props.onEdit();
    this.setState({ editText: this.props.todo.title });
  }

  handleKeyDown(event) {
    if (event.which === ESCAPE_KEY) {
      this.setState({ editText: this.props.todo.title });
      this.props.onCancel(event);
    } else if (event.which === ENTER_KEY) {
      this.handleSubmit(event);
    }
  }

  handleChange(event) {
    if (this.props.editing) {
      this.setState({ editText: event.target.value });
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    return (
      nextProps.todo !== this.props.todo ||
      nextProps.editing !== this.props.editing ||
      nextState.editText !== this.state.editText
    );
  }
  componentDidUpdate(prevProps) {
    if (!prevProps.editing && this.props.editing) {
      var node = React.findDOMNode(this.refs.editField);
      node.focus();
      node.setSelectionRange(node.value.length, node.value.length);
    }
  }

  render() {
    return (
      <li
        className={{
          completed: this.props.todo.completed,
          editing: this.props.editing
        }}
      >
        <div className="view">
          <input
            className="toggle"
            type="checkbox"
            checked={this.props.todo.completed}
            onChange={this.props.onToggle}
          />
          <label onDoubleClick={this.handleEdit}>{this.props.todo.title}</label>
          <button className="destroy" onClick={this.props.onDestroy} />
        </div>
        <input
          ref="editField"
          className="edit"
          value={this.props.editText}
          onBlur={this.handleSubmit}
          onChange={this.handleChange}
          onKeyDown={this.handleKeyDown}
        />
      </li>
    );
  }
}
