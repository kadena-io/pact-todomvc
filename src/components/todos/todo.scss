$complete: #900;
$correct: #09c;

.todo {
  display: flex;
  margin-bottom: 0.5em;

  button {
    appearance: none;
    padding: 0.5em;
    border-radius: 3px;
    color: #333;
    background: #fff;
    border: none;
    cursor: pointer;
  }

  .check button:hover {
    background: #09c;
    color: #fff;
  }

  .remove button:hover {
    background: #c00;
    color: #fff;
  }

  .entry {
    flex: 1;
    text-align: left;
    margin: 0 0.2em;

    input {
      color: #000;
      border: 1px solid transparent;
      transition: all 0.1s ease-in-out;
      border-radius: 3px;
      width: 100%;
      height: 100%;
      padding: 0 0.5em;
      font-size: 0.9em;

      &:hover,
      &.empty {
        border-color: #ccc;
      }

      &:focus {
        border-color: #999;
      }
    }
  }

  &.new {
    .entry {
      height: 1.5em;
    }
  }
  &.complete {
    color: $complete;
    .entry {
      text-decoration: strikethrough;
    }
  }
}
