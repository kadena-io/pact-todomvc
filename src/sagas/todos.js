import { call, put } from 'redux-saga/effects';

import sendPactCommand from '../utils/send-pact-command';

export const FETCH_TODOS = 'pact-todomvc/todos/FETCH_TODOS';
export const FETCH_TODOS_REQUEST = 'pact-todomvc/todos/FETCH_TODOS_REQUEST';
export const FETCH_TODOS_SUCCEEDED = 'pact-todomvc/todos/FETCH_TODOS_SUCCEEDED';
export const FETCH_TODOS_FAILED = 'pact-todomvc/todos/FETCH_TODOS_FAILED';

export const SAVE_NEW_TODO = 'pact-todomvc/todos/SAVE_NEW_TODO';
export const SAVE_NEW_TODO_REQUEST = 'pact-todomvc/todos/SAVE_NEW_TODO_REQUEST';
export const SAVE_NEW_TODO_SUCCEEDED = 'pact-todomvc/todos/SAVE_NEW_TODO_SUCCEEDED';
export const SAVE_NEW_TODO_FAILED = 'pact-todomvc/todos/SAVE_NEW_TODO_FAILED';

export const REMOVE_TODO = 'pact-todomvc/todos/REMOVE_TODO';
export const REMOVE_TODO_REQUEST = 'pact-todomvc/todos/REMOVE_TODO_REQUEST';
export const REMOVE_TODO_SUCCEEDED = 'pact-todomvc/todos/REMOVE_TODO_SUCCEEDED';
export const REMOVE_TODO_FAILED = 'pact-todomvc/todos/REMOVE_TODO_FAILED';

export const UPDATE_TODO = 'pact-todomvc/todos/UPDATE_TODO';
export const UPDATE_TODO_REQUEST = 'pact-todomvc/todos/UPDATE_TODO_REQUEST';
export const UPDATE_TODO_SUCCEEDED = 'pact-todomvc/todos/UPDATE_TODO_SUCCEEDED';
export const UPDATE_TODO_FAILED = 'pact-todomvc/todos/UPDATE_TODO_FAILED';

export const TOGGLE_TODO_STATE = 'pact-todomvc/todos/TOGGLE_TODO_STATE';
export const TOGGLE_TODO_STATE_REQUEST = 'pact-todomvc/todos/TOGGLE_TODO_STATE_REQUEST';
export const TOGGLE_TODO_STATE_SUCCEEDED = 'pact-todomvc/todos/TOGGLE_TODO_STATE_SUCCEEDED';
export const TOGGLE_TODO_STATE_FAILED = 'pact-todomvc/todos/TOGGLE_TODO_STATE_FAILED';

export const CHANGE_EDIT_STATUS = 'pact-todomvc/todos/CHANGE_EDIT_STATUS';
export const CHANGE_TODO_ENTRY = 'pact-todomvc/todos/CHANGE_TODO_ENTRY';
export const UPDATE_NEW_TODO_FIELD = 'pact-todomvc/todos/UPDATE_NEW_TODO_FIELD';

const initialState = {
  todosIsLoading: false,
  todosError: null,
  todos: [],
  newTodo: '',
  editStatus: false,
  editedTodo: null,
};

let workingTodo;
export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case FETCH_TODOS_REQUEST:
      return { ...state, todosIsLoading: true };

    case FETCH_TODOS_SUCCEEDED:
      return {
        ...state,
        todosError: null,
        todos: action.todos
          .sort((a, b) => a.id - b.id)
          .map(todo => ({ ...todo, editStatus: false })),
        todosIsLoading: false,
      };

    case FETCH_TODOS_FAILED:
      return {
        ...state,
        todosError: action.error,
        todos: [],
        todosIsLoading: false,
      };

    case SAVE_NEW_TODO_REQUEST:
      return { ...state, todosIsLoading: true };

    case SAVE_NEW_TODO_SUCCEEDED:
      return {
        ...state,
        todosError: null,
        todos: [...state.todos, action.newTodo].sort((a, b) => a.id - b.id),
        todosIsLoading: false,
        newTodo: '',
      };

    case SAVE_NEW_TODO_FAILED:
      return {
        ...state,
        todosError: action.error,
        todosIsLoading: false,
      };

    case REMOVE_TODO_REQUEST:
      return { ...state, todosIsLoading: true };

    case REMOVE_TODO_SUCCEEDED:
      return {
        ...state,
        todosError: null,
        todos: state.todos.filter(todo => todo.id !== action.id),
        todosIsLoading: false,
        newTodo: '',
      };

    case REMOVE_TODO_FAILED:
      return {
        ...state,
        todosError: action.error,
        todosIsLoading: false,
      };

    case UPDATE_TODO_REQUEST:
      return { ...state, todosIsLoading: true };

    case UPDATE_TODO_SUCCEEDED:
      return {
        ...state,
        todosError: null,
        todos: [
          ...state.todos.filter(todo => todo.id !== action.todo.id),
          { ...action.todo, editStatus: false },
        ].sort((a, b) => a.id - b.id),
        todosIsLoading: false,
      };

    case UPDATE_TODO_FAILED:
      return {
        ...state,
        todosError: action.error,
        todosIsLoading: false,
      };

    case TOGGLE_TODO_STATE_REQUEST:
      return { ...state, todosIsLoading: true };

    case TOGGLE_TODO_STATE_SUCCEEDED:
      workingTodo = state.todos.find(todo => todo.id === action.id);
      workingTodo.state = action.state;
      return {
        ...state,
        todos: [...state.todos.filter(todo => todo.id !== action.id), workingTodo].sort(
          (a, b) => a.id - b.id
        ),
        todosError: null,
        todosIsLoading: false,
      };

    case TOGGLE_TODO_STATE_FAILED:
      return {
        ...state,
        todosError: action.error,
        todosIsLoading: false,
      };

    case CHANGE_TODO_ENTRY:
      workingTodo = state.todos.find(todo => todo.id === action.id);
      workingTodo.entry = action.entry;
      return {
        ...state,
        todos: [...state.todos.filter(todo => todo.id !== action.id), workingTodo].sort(
          (a, b) => a.id - b.id
        ),
      };

    case CHANGE_EDIT_STATUS:
      workingTodo = state.todos.find(todo => todo.id === action.id);
      workingTodo.editStatus = !workingTodo.editStatus;
      return {
        ...state,
        todos: [...state.todos.filter(todo => todo.id !== action.id), workingTodo].sort(
          (a, b) => a.id - b.id
        ),
      };

    case UPDATE_NEW_TODO_FIELD:
      return { ...state, newTodo: action.newTodo };

    default:
      return state;
  }
}

export function updateNewTodoField(newTodo) {
  return { type: UPDATE_NEW_TODO_FIELD, newTodo };
}

export function fetchTodos() {
  return { type: FETCH_TODOS };
}

export function saveNewTodo(entry, date) {
  return { type: SAVE_NEW_TODO, entry, date };
}

export function changeEntry(id, entry, date) {
  return {
    type: CHANGE_TODO_ENTRY,
    id,
    entry,
    date,
  };
}

export function updateTodo(todo) {
  return { type: UPDATE_TODO, todo };
}

export function toggleState(id, state) {
  return { type: TOGGLE_TODO_STATE, id, state };
}

export function removeTodo(id) {
  return { type: REMOVE_TODO, id };
}

export function changeEditStatus(id) {
  return { type: CHANGE_EDIT_STATUS, id };
}

export function* fetchTodosSaga() {
  yield put({ type: FETCH_TODOS_REQUEST });
  try {
    const todos = yield call(sendPactCommand, '(todos.read-todos)');
    yield put({ type: FETCH_TODOS_SUCCEEDED, todos });
  } catch (error) {
    yield put({ type: FETCH_TODOS_FAILED, error });
  }
}

export function* toggleTodoStateSaga({ id, state }) {
  yield put({ type: TOGGLE_TODO_STATE_REQUEST });
  try {
    yield call(sendPactCommand, `(todos.toggle-todo-status ${id})`);
    yield put({ type: TOGGLE_TODO_STATE_SUCCEEDED, id, state });
  } catch (error) {
    yield put({ type: TOGGLE_TODO_STATE_FAILED, error });
  }
}

export function* saveNewTodoSaga({ entry, date }) {
  yield put({ type: SAVE_NEW_TODO_REQUEST });
  try {
    const newTodo = yield call(
      sendPactCommand,
      `(todos.new-todo ${JSON.stringify(entry)} ${JSON.stringify(date)})`
    );
    yield put({ type: SAVE_NEW_TODO_SUCCEEDED, newTodo });
  } catch (error) {
    yield put({ type: SAVE_NEW_TODO_FAILED, error });
  }
}

export function* removeTodoSaga({ id }) {
  yield put({ type: REMOVE_TODO_REQUEST, id });
  try {
    yield call(sendPactCommand, `(todos.delete-todo ${id})`);
    yield put({ type: REMOVE_TODO_SUCCEEDED, id });
  } catch (error) {
    yield put({ type: REMOVE_TODO_FAILED, error });
  }
}

export function* updateTodoSaga({ todo }) {
  yield put({ type: UPDATE_TODO_REQUEST, todo });
  try {
    yield call(
      sendPactCommand,
      `(todos.edit-todo ${todo.id} ${JSON.stringify(todo.entry)}  ${JSON.stringify(todo.date)})`
    );
    yield put({ type: UPDATE_TODO_SUCCEEDED, todo });
  } catch (error) {
    yield put({ type: UPDATE_TODO_FAILED, error });
  }
}
