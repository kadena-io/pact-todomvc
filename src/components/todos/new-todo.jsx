import * as React from 'react';
import './todo.scss';

export const NewTodo = ({ saveNewTodo }) => {
  let entry='';
  let date;

  const handleDate = e => {
    e.preventDefault();
    date = e.currentTarget.value;
  };

  const handleEntry = e => {
    e.preventDefault();
    entry = e.currentTarget.value;
  };

  const entryKeyDown = e => {
    if (e.keyCode === 13) {
      let newDate = new Date();
      date = date || newDate.toISOString().slice(0,10);
      console.log(entry, date)
      saveNewTodo(entry, date);
      entry='';
    }
  };

  return (
    <div className="todo new">
      <div className="entry">
        <label>New Task</label>
        <input id= "new-todo" type="text" className="empty" onChange={handleEntry} onKeyDown={entryKeyDown} placeholder="New Todoâ€¦" />
      </div>
      <div className="date">
        <label>Due Date</label>
        <input id="due-date" type="date" onChange={handleDate} onKeyDown={entryKeyDown} placeholder="Enter Due Date..." />
      </div>
    </div>
  );
};
