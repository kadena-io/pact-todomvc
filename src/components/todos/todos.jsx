import * as React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import Loader from 'react-loader-spinner';

import { fetchTodos } from '../../sagas/todos';

import { Todo } from './todo';

class TodosComponent extends React.PureComponent {
  constructor() {
    super();
    this.fetchTodos = this.fetchTodos.bind(this);
  }

  componentDidMount() {
    this.fetchTodos();
  }

  fetchTodos() {
    this.props.fetchTodos();
  }

  render() {
    const { todos, todosIsLoading, todosError } = this.props;

    if (todosIsLoading) {
      return (
        <div className="todos">
          <h1>Todos</h1>
          <Loader type="TailSpin" color="#3cf" />
        </div>
      );
    } else if (todosError !== null) {
      return (
        <div className="todos">
          <h1>Todos</h1>
          <div className="error">Error loading todos</div>
          <p>
            <Link className="btn" to="/" onClick={this.fetchTodos}>
              Reload
            </Link>
          </p>
        </div>
      );
    } else {
      return (
        <div className="todos">
          <h1>Todos</h1>'
          <div className="show-completed">
            <input type="checkbox" /> Show Completed Items
          </div>
          <div className="todos-list">
            {todos.map((todo, i) => (
              <Todo {...todo} key={i} />
            ))}
          </div>
          <p>
            <Link className="btn" to="/">
              New Todo
            </Link>
          </p>
        </div>
      );
    }
  }
}

const mapStateToProps = state => {
  /*
    todosIsLoading: false,
    todosError: null,
    todos: []
  */
  return { ...state.todos };
};

const mapDispatchToProps = dispatch => {
  return {
    fetchTodos: () => {
      dispatch(fetchTodos());
    },
    saveNewTodo: newTodo => {
      dispatch(saveNewTodo(newTodo));
    },
    removeTodo: id => {
      dispatch(removeTodo(id));
    },
  };
};

export const Todos = connect(
  mapStateToProps,
  mapDispatchToProps
)(TodosComponent);
