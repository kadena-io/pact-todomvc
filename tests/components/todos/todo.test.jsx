import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, render } from 'enzyme';

import { Todo } from '../../../src/components/todos/todo';

const todoProps = {
  id: 1,
  entry: 'A Todo',
  state: 'active',
  date: '2018-10-09',
  deleted: false,
  onChangeEntry: jest.fn(),
  onChangeDate: jest.fn(),
  onUpdate: jest.fn(),
  onRemove: jest.fn(),
  onToggleState: jest.fn(),
  onClickEdit: jest.fn(),
};

describe('Todo Component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should render a todo', () => {
    const component = renderer.create(<Todo {...todoProps} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should correctly render icon based on state prop', () => {
    const active = render(<Todo {...todoProps} state="active" />);
    const activeIcon = active.find('.check button svg');
    expect(active.hasClass('active')).toBeTruthy();
    expect(activeIcon.prop('data-prefix')).toEqual('far');

    const completed = render(<Todo {...todoProps} state="completed" />);
    const completedIcon = completed.find('.check button svg');
    expect(completed.hasClass('completed')).toBeTruthy();
    expect(completedIcon.prop('data-prefix')).toEqual('fas');
  });

  test('should correctly set class on entry input based on entry length', () => {
    const filled = render(<Todo {...todoProps} entry="Some Text" editStatus={true} />);
    const filledInput = filled.find('.entry input');
    expect(filledInput.hasClass('empty')).toBeFalsy();

    const empty = render(<Todo {...todoProps} entry="" editStatus={true} />);
    const emptyInput = empty.find('.entry input');
    expect(emptyInput.hasClass('empty')).toBeTruthy();
  });

  test('should call onRemove() when clicking remove button', () => {
    const component = shallow(<Todo {...todoProps} />);
    const removeButton = component.find('.remove button');
    removeButton.simulate('click');
    expect(todoProps.onRemove).toHaveBeenCalled();
  });

  test('should call onToggleState() when clicking complete button button', () => {
    const active = shallow(<Todo {...todoProps} state="active" />);
    const activeButton = active.find('.check button');
    activeButton.simulate('click');
    expect(todoProps.onToggleState).toHaveBeenCalledWith(todoProps.id, 'completed');

    todoProps.onToggleState.mockClear();

    const completed = shallow(<Todo {...todoProps} state="completed" />);
    const completedButton = completed.find('.check button');

    completedButton.simulate('click');
    expect(todoProps.onToggleState).toHaveBeenCalledWith(todoProps.id, 'active');
  });

  test('should call changeEntry() when changing entry field', () => {
    const component = shallow(<Todo {...todoProps} editStatus={true} />);
    const input = component.find('.entry input');
    const newValue = 'Something Else';
    input.simulate('change', { currentTarget: { value: newValue } });
    expect(todoProps.onChangeEntry).toHaveBeenCalledWith(todoProps.id, newValue);
  });

  test('should call changeDate() when changing entry field', () => {
    const component = shallow(<Todo {...todoProps} editStatus={true} />);
    const input = component.find('.due-date input');
    const newValue = '2017-10-01';
    input.simulate('change', { currentTarget: { value: newValue } });
    expect(todoProps.onChangeDate).toHaveBeenCalledWith(todoProps.id, newValue);
  });

  test('should call onclickEdit() when edit button is clicked', () => {
    const component = shallow(<Todo {...todoProps} editStatus={true} />);
    const editButton = component.find('.edit button');
    editButton.simulate('click');
    expect(todoProps.onClickEdit).toHaveBeenCalledWith(todoProps.id);
  });

  test('should call onUpdate() when blurring date field', () => {
    const component = shallow(<Todo {...todoProps} editStatus={true} />);
    component.find('.due-date input').simulate('blur');
    expect(todoProps.onUpdate).toHaveBeenCalledWith({
      id: todoProps.id,
      entry: todoProps.entry,
      date: todoProps.date,
      state: todoProps.state,
      deleted: todoProps.deleted,
    });
  });

  test('should call onUpdate() when blurring entry field', () => {
    const component = shallow(<Todo {...todoProps} editStatus={true} />);
    component.find('.entry input').simulate('blur');
    expect(todoProps.onUpdate).toHaveBeenCalledWith({
      id: todoProps.id,
      entry: todoProps.entry,
      date: todoProps.date,
      state: todoProps.state,
      deleted: todoProps.deleted,
    });
  });

  test('should only call blur() on text field if keyDown keyCode is 13', () => {
    const component = shallow(<Todo {...todoProps} editStatus={true} />);
    const input = component.find('.entry input');
    const newValue = 'Something Else';
    const currentTarget = { blur: jest.fn() };

    input.simulate('keyDown', { keyCode: 1, currentTarget });
    expect(currentTarget.blur).not.toHaveBeenCalled();

    input.simulate('keyDown', { keyCode: 13, currentTarget });
    expect(currentTarget.blur).toHaveBeenCalled();
  });
});
