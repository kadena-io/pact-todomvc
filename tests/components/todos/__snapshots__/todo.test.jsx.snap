// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Component should render a todo 1`] = `
<div
  className="todo active"
>
  <div
    className="check"
  >
    <button
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-circle fa-w-16 "
        data-icon="circle"
        data-prefix="far"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </button>
  </div>
  <div
    className="entry"
  >
    <span>
      A Todo
    </span>
  </div>
  <div
    className="due-date"
  >
    <span>
      10-09-2018
    </span>
  </div>
  <div
    className="edit"
  >
    <button
      disabled={false}
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-edit fa-w-18 "
        data-icon="edit"
        data-prefix="fas"
        role="img"
        style={Object {}}
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </button>
  </div>
  <div
    className="remove"
  >
    <button
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-times fa-w-11 "
        data-icon="times"
        data-prefix="fas"
        role="img"
        style={Object {}}
        viewBox="0 0 352 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </button>
  </div>
</div>
`;
